Fixes zone/region labels setup and kubelet stucking on startup if credentials stored in secret for legacy vSphere cloudprovider.

lobziik opened the pull request:
#### What this PR does / why we need it:
Disables attempts of obtaining zones within kubelet during initial node registration by vSphere provider if credentials stored in secret.
Setting zone and region labels for node moved to KCM in such case.

If credentials stored in cloud-provider config file as plaintext current behaviour does not change.

#### Which issue(s) this PR fixes:
Fixes #75175. Kubelet does not stucking on startup with this patch. Zone labels populates for nodes during KCM startup.

#### Notes:
For proper functioning ClusterRole + ClusterRoleBinding need to be created if RBAC is in use. This should be documented somewhere, would be awesome if anybody will point me a good place for this. In case of lack of permissions labels will not be set.

#### What type of PR is this?
/kind bug
/sig cloud-provider

/assing @andrewsykim

#### Does this PR introduce a user-facing change?
```release-note
Fixed bug with leads to Node goes "Not-ready" state when credentials for vCenter stored in a secret and Zones feature is in use.
Zone labels setup moved to KCM component, kubelet skips this step during startup in such case. If credentials stored in cloud-provider config file as plaintext current behaviour does not change and no action required.

Action required: For proper functioning "kube-system:vsphere-legacy-cloud-provider" should be allowed to update node object if vCenter credentials stored in secret and Zone feature used.
```

Commit:
Set zones for node within kcm if creds stored in secret

Commit:
Disable zones obtaining attempts for legacy vSphere cloud provider if secret provided and no CredentialsManager was set up.

Partially solves #75175. Kubelet does not stucking on startup.

Commit:
Commonize GetZone with GetZoneByProviderID, retries for node update in case of conflicts

Commit:
reordered imports

Danil-Grigorev commented:
/ok-to-test

yangjunmyfm192085 commented on staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:
```diff
@@ -894,7 +901,12 @@ func (vs *VSphere) LoadBalancer() (cloudprovider.LoadBalancer, bool) {
-	return vs.cfg != nil && vs.cfg.Labels.Zone != "" && vs.cfg.Labels.Region != ""
+	isEnabled := vs.cfg != nil && vs.cfg.Labels.Zone != "" && vs.cfg.Labels.Region != ""
+	if isEnabled && vs.isSecretInfoProvided && vs.nodeManager.credentialManager == nil {
+		klog.V(1).Info("Zones can not be populated now due to credentials in Secret, skip.")
```
```suggestion
		klog.V(1).InfoS("Zones can not be populated now due to credentials in Secret, skip.")
```

yangjunmyfm192085 commented on staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:
```diff
@@ -1519,6 +1531,51 @@ func (vs *VSphere) NodeAdded(obj interface{}) {
+	if vs.isSecretInfoProvided && vs.isZoneEnabled() {
+		zone, err := vs.GetZoneByProviderID(ctx, node.Spec.ProviderID)
+		if err != nil {
+			klog.Warningf("Can not get Zones from vCenter: %v", err)
```
```suggestion
			klog.InfoS("Can not get Zones from vCenter", "err", err)
```

cheftako replied:
Not sure we should do this upgrade piecemeal. Seems like this should be done for the whole section.

yangjunmyfm192085 commented on staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:
```diff
@@ -1519,6 +1531,51 @@ func (vs *VSphere) NodeAdded(obj interface{}) {
+
+			err = tryUpdateNode(ctx, vs.kubeClient, updatedNode)
+			if err != nil {
+				klog.Errorf("vSphere cloud provider can not update node with zones info: %v", err)
```
```suggestion
				klog.InfoS("VSphere cloud provider can not update node with zones info", "err", err)
```

yangjunmyfm192085 commented on staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:
```diff
@@ -1519,6 +1531,51 @@ func (vs *VSphere) NodeAdded(obj interface{}) {
+			if err != nil {
+				klog.Errorf("vSphere cloud provider can not update node with zones info: %v", err)
+			} else {
+				klog.V(4).Infof("Node %s updated with zone and region labels", updatedNode.Name)
```
```suggestion
				klog.V(4).InfoS("Node updated with zone and region labels", "nodeName", updatedNode.Name)
```

yangjunmyfm192085 commented on staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:
```diff
@@ -1701,12 +1753,31 @@ func (vs *VSphere) GetZone(ctx context.Context) (cloudprovider.Zone, error) {
+func (vs *VSphere) GetZone(ctx context.Context) (cloudprovider.Zone, error) {
+	nodeName, err := vs.CurrentNodeName(ctx, vs.hostName)
+	if err != nil {
+		klog.Errorf("Cannot get node name.")
```
```suggestion
		klog.ErrorS(nil, "Cannot get node name.")
```

yangjunmyfm192085 commented:
Is it necessary to  structured log migration?

lobziik commented:
@yangjunmyfm192085 Dunno, in my opinion logs format migration should be done for entire module. Would prefer to keep it consistent with other provider's code for now.

lobziik commented:
/retest

pacoxu commented on staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere_test.go:
```diff
@@ -371,6 +371,62 @@ func TestZonesNoConfig(t *testing.T) {
 }
 
+func TestZonesWithCredsInSecret(t *testing.T) {
+	noSecretCfg, err := readConfig(strings.NewReader(`
```
what if err is not nil here.

pacoxu commented on staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere_test.go:
```diff
@@ -371,6 +371,62 @@ func TestZonesNoConfig(t *testing.T) {
+	// Such situation happens during kubelet startup process, when InitialNode creates.
+	// See https://github.com/kubernetes/kubernetes/issues/75175
+	// and https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L418
+	withSecretCfg, err := readConfig(strings.NewReader(`
```
same question here. not used `err`.

Danil-Grigorev commented on staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:
```diff
@@ -894,7 +901,12 @@ func (vs *VSphere) LoadBalancer() (cloudprovider.LoadBalancer, bool) {
 func (vs *VSphere) isZoneEnabled() bool {
-	return vs.cfg != nil && vs.cfg.Labels.Zone != "" && vs.cfg.Labels.Region != ""
+	isEnabled := vs.cfg != nil && vs.cfg.Labels.Zone != "" && vs.cfg.Labels.Region != ""
+	if isEnabled && vs.isSecretInfoProvided && vs.nodeManager.credentialManager == nil {
```
Why is this exactly like that? Should not informers and credentialManager eventually set up and stop being nil? Because returning false here is permanent - no other consumer will be able to initialize node with this labels

Danil-Grigorev replied:
Please add a comment specifying that such combination of values when `vs.isSecretInfoProvided && vs.nodeManager.credentialManager != nil` is unique to KCM only. So that it will be clear this code will return false only in kubelet.

lobziik replied:
Kubelet will never have this `credentialManager`, informers doesn't setting up there. This thing just for workaround it and prevent kubelet from hanging up during startup if credentials stored in secret.

Within KCM or with plain text credentials all works as before.

Danil-Grigorev commented on staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:
```diff
@@ -1519,6 +1531,51 @@ func (vs *VSphere) NodeAdded(obj interface{}) {
 	if err := vs.nodeManager.RegisterNode(node); err != nil {
 		klog.Errorf("failed to add node %+v: %v", node, err)
 	}
+	vs.setNodeZoneLabels(node)
```
This appears to be the only place where the labels are set, and it is called only once for each node object. Wouldn't it result in nodes without zone labels at all in a cluster where the labels should be set?

lobziik replied:
Shouldn't. This thing calls when kcm recognizes new node and adds it to internal nodeManager structure.

Commit:
add description for elaborate changes in isZonesEnabled method

lobziik commented:
@andrewsykim @cheftako, Hello! Could you please take a brief look on this? Is this approach reasonable? :)

cheftako commented on staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:
```diff
@@ -894,7 +901,16 @@ func (vs *VSphere) LoadBalancer() (cloudprovider.LoadBalancer, bool) {
 
 func (vs *VSphere) isZoneEnabled() bool {
-	return vs.cfg != nil && vs.cfg.Labels.Zone != "" && vs.cfg.Labels.Region != ""
+	isEnabled := vs.cfg != nil && vs.cfg.Labels.Zone != "" && vs.cfg.Labels.Region != ""
```
Is it really both zone and region need to be set or just one of zone and region needs to be set?

lobziik replied:
https://vmware.github.io/vsphere-storage-for-kubernetes/documentation/zones.html this doc tells that both need to be set. Case with one parameter does not described at all.

cheftako commented:
Happy to approve but would like to see one of the primary vsphere contributors lgtm first.

lobziik commented:
@divyenpatel @andrewsykim @sandeeppissay Hello! Could you please take a look on this please? Would be awesome to get a feedback from vsphere maintainers :)

SandeepPissay commented:
@lobziik unfortunately I do not have bandwidth to review code changes in legacy vSphere cloud provider. Its also been a while I looked at that code.

jcpowermac commented:
> @lobziik unfortunately I do not have bandwidth to review code changes in legacy vSphere cloud provider. Its also been a while I looked at that code.

@SandeepPissay our mutal customers would like to be able to support zones. If you are unable to review is there someone else in the reviewer/approver list that can?

cheftako commented:
/approve

JoelSpeed commented:
Code wise, this LGTM, but I'm not a vSphere maintainer or expert, just a user. I could add the relevant label but I appreciate that this would cause the code to merge.

I might suggest, if we can't get anyone who knows the vSphere code better, to set a lazy consensus, any objections?

dims commented:
cc @randomvariable

lubronzhan commented on staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:
```diff
@@ -894,7 +901,16 @@ func (vs *VSphere) LoadBalancer() (cloudprovider.LoadBalancer, bool) {
+	// Return false within kubelet in case of credentials stored in secret.
+	// Otherwise kubelet will not be able to obtain zone labels from vSphere and create initial node
+	// due to no credentials at this step.
+	// See https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L394 for more details.
```
Could you pin this to a commit, since master branch could be changed. I think you are referring to this part https://github.com/kubernetes/kubernetes/blob/bad4faf1b9d69d3cb6e8de90dc5eee1d8cf0b6d3/pkg/kubelet/kubelet_node_status.go#L384-L386

lubronzhan commented on staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:
```diff
@@ -1519,6 +1535,51 @@ func (vs *VSphere) NodeAdded(obj interface{}) {
+			if !apierrors.IsConflict(err) {
+				return fmt.Errorf("vSphere cloud provider can not update node with zones info: %v", err)
+			}
+		} else {
```
nit: else is not needed

lubronzhan commented on staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:
```diff
@@ -1701,12 +1757,31 @@ func (vs *VSphere) GetZone(ctx context.Context) (cloudprovider.Zone, error) {
+	for nName, nInfo := range vs.nodeManager.nodeInfoMap {
+		if nInfo.vmUUID == vmUUID {
+			nodeName = convertToK8sType(nName)
+		}
```
nit: could add a break after finding break

lubronzhan commented:
LGTM. @yastij  could you also take a look?

andrewsykim commented:
> Action required: For proper functioning "kube-syste:vsphere-legacy-cloud-provider" should be allowed to update node object if vCenter credentials stored in secret and Zone feature used.

I noticed this in the release notes, is `kube-syste:vsphere-legacy-cloud-provider` an official role or just an example? We should make this clear in the release notes

lobziik commented:
Updated pr a bit, what was done:
- rebased on top of current master
- fixes according @lubronzhan review

>I noticed this in the release notes, is kube-syste:vsphere-legacy-cloud-provider an official role or just an example? We should make this clear in the release notes

@andrewsykim, official, i think. I introduced client instantiation here: https://github.com/kubernetes/kubernetes/pull/101028/files#diff-5000d45019379218cc35bc5f967f2731b4dbff230399361c8f6b9a8af3e7e1f0R277
Name is just hardcoded as in other legacy providers. I fixed typo in PR description, thanks for catching that!

lobziik commented:
/test pull-kubernetes-integration

andrewsykim commented:
> @andrewsykim, official, i think. I introduced client instantiation here: https://github.com/kubernetes/kubernetes/pull/101028/files#diff-5000d45019379218cc35bc5f967f2731b4dbff230399361c8f6b9a8af3e7e1f0R277
Name is just hardcoded as in other legacy providers. I fixed typo in PR description, thanks for catching that!

Fwiw I think this is just used as the user-agent in the client and not for RBAC

lubronzhan commented:
Thanks for reminding Andrew! 
Just realized that. If we don't automatically add the role binding and role before node registration, the code [here](https://github.com/kubernetes/kubernetes/pull/101028/files#diff-5000d45019379218cc35bc5f967f2731b4dbff230399361c8f6b9a8af3e7e1f0R1544) will try send 3 requests to APIserver and get error of not enough permission, failing to add the labels and won't ever add it again.
Unless customer add those rolebinding and role, then recreate nodes again.
Should we consider adding the rolebinding and role as default?

lubronzhan commented:
Looking at the code of ClientOrDie here, looks like the input will be used as service account name to create config
https://github.com/kubernetes/kubernetes/blob/bfdbfa2ab6f1710077fe1c44b5fa3a21949b7c39/staging/src/k8s.io/controller-manager/pkg/clientbuilder/client_builder_dynamic.go#L121-L129

andrewsykim commented:
Ah you're right @lubronzhan, I was thinking of the simple client builder:

https://github.com/kubernetes/kubernetes/blob/bfdbfa2ab6f1710077fe1c44b5fa3a21949b7c39/staging/src/k8s.io/controller-manager/pkg/clientbuilder/client_builder.go#L44-L48

andrewsykim commented:
> Should we consider adding the rolebinding and role as default?

We no longer support bootstrapping role/rolebindings specific to cloud providers, these should be handled externally

lubronzhan commented:
If the role/rolebinding is handled externally, could we run into this issue?
> If we don't automatically add the role binding and role before node registration, the code here will try send 3 requests to APIserver and get error of not enough permission, failing to add the labels and won't ever add it again.
>Unless customer add those rolebinding and role, then recreate nodes again.

lobziik commented:
@lubronzhan,  I went through the code again and yes, seems this will engage only once - during new Node handling within KCM...

I will need check this again from kubelet perspective, poked around quite a while ago and don't exactly remember how it behaves with secret there.

lobziik commented:
I dug into kubelet code, and, as far as I can say there is only one place where zone labels populates:
https://github.com/kubernetes/kubernetes/blob/bad4faf1b9d69d3cb6e8de90dc5eee1d8cf0b6d3/pkg/kubelet/kubelet_node_status.go#L406

In other words, yes, in case of not enough permissions this what would happen exactly.
>If we don't automatically add the role binding and role before node registration, the code here will try send 3 requests to APIserver and get error of not enough permission, failing to add the labels and won't ever add it again.
Unless customer add those rolebinding and role, then recreate nodes again.

lobziik commented:
I attempted to fix the case with not sufficient permissions in https://github.com/kubernetes/kubernetes/pull/101028/commits/3dfc0113a9fe5cc1154b043d30074837a460e258

@lubronzhan could you please take a look?

lubronzhan commented:
Hi @lobziik Right now the node label will be set every time when node is updated.
Does that mean after the customer add the role and rolebinding he needs to somehow trigger the node updates?

lobziik commented:
>Hi @lobziik Right now the node label will be set every time when node is updated.
Does that mean after the customer add the role and rolebinding he needs to somehow trigger the node updates?

Hi! For some reason i thought that this handler would be triggered time to time, during status update/heartbeat, but seems not. So, yes, in current variant node update need to be to be triggered somehow for reconcile zone labels.

lobziik committed:
Add zone/region labels sync function

In the case of `vsphere-legacy-cloud-provider` client
has insufficient permissions to update Node after its addition,
this handler will attempt to populate Nodes topology labels later.

lobziik commented:
Gentle ping @lubronzhan, @andrewsykim :)

lubronzhan commented:
Sorry I missed the email notification.
Yeah now it makes more sense!
/approve
/lgtm
